tests/shrub/v3/test_shrub_service.py

```diff
@@ -0,0 +1,139 @@
+"""Unit tests for shrub_service.py."""
+
+import json
+import pytest
+
+from shrub.v3.evg_task import EvgTask, EvgTaskDependency
+from shrub.v3.evg_build_variant import BuildVariant, DisplayTask
+from shrub.v3.evg_command import FunctionCall, shell_exec
+from shrub.v3.evg_project import EvgProject
+from shrub.v3.shrub_service import ShrubService
+
+
+@pytest.fixture
+def project():
+    n_tasks = 10
+
+    def define_task(index):
+        name = f"task_name_{index}"
+
+        return EvgTask(
+            name=name,
+            commands=[
+                FunctionCall(func="do setup"),
+                FunctionCall(
+                    func="run test generator",
+                    vars={"parameter_1": "value 1", "parameter_2": "value 2"},
+                ),
+                FunctionCall(func="run tests"),
+                shell_exec("a=1\nb=2", env=dict(foo="true", bar="1", fizz=True, buzz=1)),
+            ],
+            tags=[f"tag{i}" for i in range(index)],
+            depends_on=[EvgTaskDependency(name="compile")],
+        )
+
+    tasks = [define_task(i) for i in range(n_tasks)]
+    display_task = DisplayTask(name="test_suite", execution_tasks=[t.name for t in tasks])
+    variant = BuildVariant(name="linux-64", tasks=[], display_tasks=[display_task])
+    return EvgProject(buildvariants=[variant], tasks=tasks)
+
+
+def test_project_json(project):
+    out = ShrubService.generate_json(project)
+    data = json.loads(out)
+    assert "buildvariants" in data
+    assert "tasks" in data
+
+
+EXPECTED_YAML1 = """
+buildvariants:
+  - name: linux-64
+    tasks: []
+    display_tasks:
+      - name: test_suite
+        execution_tasks:
+          - task_name_0
+          - task_name_1
+          - task_name_2
+          - task_name_3
+          - task_name_4
+          - task_name_5
+          - task_name_6
+          - task_name_7
+          - task_name_8
+          - task_name_9
+tasks:
+  - name: task_name_0
+    commands:
+      - func: do setup
+      - func: run test generator
+        vars:
+          parameter_1: value 1
+          parameter_2: value 2
+      - func: run tests
+      - command: shell.exec
+        params:
+          script: |-
```

Comment:
```suggestion
          script: |
```
I believe this should be `|` (clip), not `|-` (strip), per `represent_scalar()`'s `style = "|"` for multiline strings. Same for other multiline strings below.

Reply:
This is the generated output, perhaps something is missing in `ConfigDumper`?

Reply:
Interesting. For `"a=1\nb=2", it emits `|-`, but for `"a=1\nb=2\n"`, it emits `|`. Seems like the dumper is smart and knows which chomping indicator to use depending on whether a trailing newline is present in the Python string or not. The currently emitted output is unexpected but nevertheless correct.

May want to add a few more string formatting tests that handle the presence of a trailing newline as well as extra newlines (which should end up being excluded).

Reply:
Updated